runtime/src/bank.rs

```diff
@@ -4788,86 +4796,114 @@ impl Bank {
         }
     }
 
-    pub fn load_program(&self, pubkey: &Pubkey) -> Arc<LoadedProgram> {
-        let Some(program) = self.get_account_with_fixed_root(pubkey) else {
-            return Arc::new(LoadedProgram::new_tombstone(
-                self.slot,
-                LoadedProgramType::Closed,
-            ));
+    fn load_program_accounts(&self, pubkey: &Pubkey) -> ProgramAccountLoadResult {
+        let program_account = match self.get_account_with_fixed_root(pubkey) {
+            None => return ProgramAccountLoadResult::AccountNotFound,
+            Some(account) => account,
         };
-        let mut transaction_accounts = vec![(*pubkey, program)];
-        let is_upgradeable_loader =
-            bpf_loader_upgradeable::check_id(transaction_accounts[0].1.owner());
-        if is_upgradeable_loader {
-            if let Ok(UpgradeableLoaderState::Program {
-                programdata_address,
-            }) = transaction_accounts[0].1.state()
+
+        if !solana_bpf_loader_program::check_loader_id(program_account.owner()) {
+            return ProgramAccountLoadResult::InvalidAccountData;
+        }
+
+        if !bpf_loader_upgradeable::check_id(program_account.owner()) {
+            return ProgramAccountLoadResult::ProgramOfLoaderV1orV2(program_account);
+        }
+
+        if let Ok(UpgradeableLoaderState::Program {
+            programdata_address,
+        }) = program_account.state()
+        {
+            let programdata_account = match self.get_account_with_fixed_root(&programdata_address) {
+                None => return ProgramAccountLoadResult::AccountNotFound,
+                Some(account) => account,
+            };
+
+            if let Ok(UpgradeableLoaderState::ProgramData {
+                slot,
+                upgrade_authority_address: _,
+            }) = programdata_account.state()
             {
-                if let Some(programdata_account) =
-                    self.get_account_with_fixed_root(&programdata_address)
-                {
-                    transaction_accounts.push((programdata_address, programdata_account));
-                } else {
-                    return Arc::new(LoadedProgram::new_tombstone(
-                        self.slot,
-                        LoadedProgramType::Closed,
-                    ));
-                }
-            } else {
-                return Arc::new(LoadedProgram::new_tombstone(
-                    self.slot,
-                    LoadedProgramType::Closed,
-                ));
+                return ProgramAccountLoadResult::ProgramOfLoaderV3(
+                    program_account,
+                    programdata_account,
+                    slot,
+                );
             }
         }
```

Comment:
There was a redundant verification in `load_program()` before `load_program_from_account()` and it returned `LoadedProgramType::Closed` on the `else` branch, so the fall through to `InvalidAccountData` is incorrect.

Reply:
That code block is replaced with this logic
```
            let programdata_account = match self.get_account_with_fixed_root(&programdata_address) {
                None => return ProgramAccountLoadResult::AccountNotFound,
                Some(account) => account,
            };
```

So the else block maps to `None`. That'll create the tombstone with `Closed` program type.

Reply:
I was talking about the case in which `programdata_account` is found but does not match `UpgradeableLoaderState::ProgramData`. Previously that was caught in `load_program_from_account()` and resulted in `InstructionError::InvalidAccountData` thus `LoadedProgramType::FailedVerification`. So it should be fine as well.